src/main/scala/com/cognite/spark/datasource/ThreeDModelsRelation.scala

```diff
@@ -1,19 +1,51 @@
 package com.cognite.spark.datasource
 
+import cats.effect.IO
+import com.codahale.metrics.Counter
+import com.cognite.sdk.scala.common
+import com.cognite.sdk.scala.v1.{GenericClient, ThreeDModel}
 import com.cognite.spark.datasource.SparkSchemaHelper._
 import com.softwaremill.sttp._
 import org.apache.spark.sql.types.{DataTypes, StructType}
 import org.apache.spark.sql.{Row, SQLContext}
 import io.circe.generic.auto._
+import org.apache.spark.datasource.MetricsSource
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.sources.{BaseRelation, TableScan}
 
 case class ModelItem(id: Long, name: String, createdTime: Long)
 
 class ThreeDModelsRelation(config: RelationConfig)(val sqlContext: SQLContext)
-    extends CdpRelation[ModelItem](config, "3dmodels") {
-  override def schema: StructType = structType[ModelItem]
+    extends BaseRelation
+    with Serializable
+    with TableScan {
+  @transient lazy protected val itemsRead: Counter =
+    MetricsSource.getOrCreateCounter(config.metricsPrefix, s"threeDModels.read")
 
-  override def toRow(t: ModelItem): Row = asRow(t)
+  override def buildScan(): RDD[Row] =
+    SdkRddV1(
+      sqlContext.sparkContext,
+      config,
+      compute,
+      cursors()
+    )
+  def schema: StructType = structType[ThreeDModel]
 
-  override def listUrl(): Uri =
-    uri"${config.baseUrl}/api/0.6/projects/${config.project}/3d/models"
+  def compute(c: GenericClient[IO, Nothing], p: SdkPartition): Iterator[Row] = {
+    if (config.collectMetrics) {
+      itemsRead.inc()
+    }
+    val items: IO[common.ItemsWithCursor[ThreeDModel]] =
+      c.threeDModels.readWithCursor(p.cursor, config.limit.map(_.toLong))
+    val it = items.unsafeRunSync()
```

Comment:
I would rather like the `unsafeRunSync` part to be at the end of `readWithCursor`

Reply:
Definitely, good catch.